helm-chart(v4): Use env var for setting CONTAINER_LOG_PATH, new default for containerRuntime (#1774)
fluent/fluent-operator: charts/fluent-operator/MIGRATION-v4.md

# Migration Guide: Fluent Operator Helm Chart v3.x to v4.0

## Overview

v4.0 simplifies container runtime configuration by removing dynamic detection for the `docker` runtime via initContainers and adopting static, configuration-based paths. The `docker` runtime has not been used widely since Kubernetes v1.24 (2022) and modern Kubernetes distributions now use the `containerd` runtime.

## Breaking Changes

### 1. Default Container Runtime Changed

**What Changed:**
- Default `containerRuntime` changed from `docker` to `containerd`

**Impact:**
- Users who never explicitly set `containerRuntime` will now use containerd defaults
- Log parser will change from `docker` to `cri` format
- Systemd filter will target `containerd.service` instead of `docker.service`

**Who Is Affected:**
- Users still running Docker container runtime (uncommon - Docker support was removed from Kubernetes in v1.24, May 2022)

> - Users who relied on the default value without explicit configuration

Reviewer: ```suggestion

- Default `containerRuntime` changed from `docker` to `containerd`

**Impact:**

- Users who never explicitly set `containerRuntime` will now use containerd defaults
- Log parser will change from `docker` to `cri` format
- Systemd filter will target `containerd.service` instead of `docker.service`

**Who Is Affected:**

- Users still running Docker container runtime (uncommon - Docker support was removed from Kubernetes in v1.24, May 2022)
- Users who relied on the default value without explicit configuration
```

Markdownlint will complain about empty lines need to be before and after unordered lists.

- initContainer no longer runs before the operator starts
- Removes dependency on third party outdated images for initContainers
- Docker socket no longer mounted for path detection

**Who Is Affected:**
- Users who customized `operator.initcontainer.image` or `operator.initcontainer.resources`
- Users with Docker installations using non-standard root directories

**Migration:**
```yaml
# v3.x configuration (REMOVED)
operator:
  initcontainer:
    image:
      registry: docker.io
      repository: docker
      tag: "20.10"
    resources:
      limits:
        cpu: 100m
        memory: 64Mi

# v4.0 - No replacement needed
# initContainer functionality is removed
# If you need custom resource limits, adjust operator.resources instead
```

### 3. Log Path Configuration Simplified

**What Changed:**
- Removed `operator.logPath.containerd` and `operator.logPath.crio`
- Added new `operator.containerLogPath` for direct path specification

**Impact:**
- Old logPath configuration is ignored
- New configuration accepts full path to container logs (not just root directory)

**Who Is Affected:**
- Users who set custom paths via `operator.logPath.containerd` or `operator.logPath.crio`

> **Migration:**

Reviewer: ```suggestion
**What Changed:**

- Removed `operator.logPath.containerd` and `operator.logPath.crio`
- Added new `operator.containerLogPath` for direct path specification

**Impact:**

- Old logPath configuration is ignored
- New configuration accepts full path to container logs (not just root directory)

**Who Is Affected:**

- Users who set custom paths via `operator.logPath.containerd` or `operator.logPath.crio`

**Migration:**

```

```yaml
# v3.x configuration (DEPRECATED)
operator:
  logPath:
    containerd: /var/log

# v4.0 - Use direct path specification
operator:
  containerLogPath: "/var/log/containers"

# Note: The v3 config specified the ROOT directory

> # The v4 config specifies the FULL path to containers directory

Reviewer: ```suggestion

In v3 the root directory was specified for the log path. In v4 this will be the full path to the container logs folder.

```diff
  operator:
-   logPath:
-     containerd: /var/log
+   containerLogPath: "/var/log/containers"
```
